lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -1710,6 +1710,47 @@ static LogicalResult verifySubfieldOp(SubfieldOp op) {
   return success();
 }
 
+/// Check that a value is driven by a constant value. Looks through subindex and
+/// subfield ops.
+static bool isDrivenByConstant(Value op) {
+  for (auto &use : op.getUses()) {
+    auto result = TypeSwitch<Operation *, bool>(use.getOwner())
+                      .Case<SubindexOp, SubfieldOp>([](auto op) {
+                        return isDrivenByConstant(op.getResult());
+                      })
+                      .Case<SubaccessOp>([](auto) { return false; })
+                      .Case<PartialConnectOp, ConnectOp>([&](auto op) {
+                        if (use.get() == op.dest())
+                          return isConstant(op.src());
+                        else
+                          return true;
+                      })
+                      .Default([](auto) { return true; });
+    if (!result)
+      return false;
+  }
+  return true;
```

Comment:
I keep writing patterns like this (myself) and wondering if they can be better expressed as an `llvm::all_of` for readability. Maybe that could be used here.

Reply:
`all_of` is a great idea!

Reply:
I reworked this to remove the recursion.  This now uses a worklist and the `all_of` idea doesn't make sense. 😅